pkg/bbgo/serialmarketdatastore.go

```diff
@@ -30,24 +43,107 @@ func (store *SerialMarketDataStore) Subscribe(interval types.Interval) {
 	store.Subscription = append(store.Subscription, interval)
 }
 
-func (store *SerialMarketDataStore) BindStream(stream types.Stream) {
-	stream.OnKLineClosed(store.handleKLineClosed)
+func (store *SerialMarketDataStore) BindStream(ctx context.Context, stream types.Stream) {
+	if store.UseMarketTrade {
+		if IsBackTesting {
+			log.Errorf("right now in backtesting, aggTrade event is not yet supported. Use OnKLineClosed instead.")
+			stream.OnKLineClosed(store.handleKLineClosed)
+			return
+		}
+		go store.tickerProcessor(ctx)
+		stream.OnMarketTrade(store.handleMarketTrade)
+	} else {
+		stream.OnKLineClosed(store.handleKLineClosed)
+	}
 }
 
 func (store *SerialMarketDataStore) handleKLineClosed(kline types.KLine) {
 	store.AddKLine(kline)
 }
 
-func (store *SerialMarketDataStore) AddKLine(kline types.KLine) {
+func (store *SerialMarketDataStore) handleMarketTrade(trade types.Trade) {
+	store.mu.Lock()
+	store.price = trade.Price
+	store.c = store.price
+	if store.price.Compare(store.h) > 0 {
+		store.h = store.price
+	}
+	if !store.l.IsZero() {
+		if store.price.Compare(store.l) < 0 {
+			store.l = store.price
+		}
+	} else {
+		store.l = store.price
+	}
+	if store.o.IsZero() {
+		store.o = store.price
+	}
+	store.v = store.v.Add(trade.Quantity)
+	store.qv = store.qv.Add(trade.QuoteQuantity)
+	store.mu.Unlock()
+}
+
+func (store *SerialMarketDataStore) tickerProcessor(ctx context.Context) {
+	duration := store.MinInterval.Duration()
+	relativeTime := time.Now().UnixNano() % int64(duration)
+	waitTime := int64(duration) - relativeTime
+	ch := time.After(time.Duration(waitTime))
+	<-ch
```

Comment:
you can just do:

```
<-time.After(time.Duration(waitTime))
```

Reply:
but it would be better to handle both ctx.Done and this channel in a select clause.